### DISCUSSION TITLE
Welcome to Orca Discussions!

### DISCUSSION THREAD

**Discussion by Soheil-ab** [Soheil-ab/Orca#7] [General]
## 👋 Welcome!
Instead of opening an issue, you can use this discussion section to:
  * Ask questions you’re wondering about.
  * Share ideas.
  * ... 

Feel free to comment below! :-)

**Comment by ZoraNeverEatFish** [comment]
About Evaluation.
The evaluation scenarios of real-world in your paper (section 5.1) are convincing, and we are trying to build a similar one to test some CC algorithms. Using Pantheon seems to be an efficient way, but we had been stuck in it (the performances within/without mm-tunnel are far different). 
We noticed that you evaluated CCs without Pantheon but using its tools (in github issue). Could you please give a brief introduction of your evaluation implement?
Thanks a lot!

**Reply by Soheil-ab** [reply]
Hey @ZoraNeverEatFish 

One important thing to consider during your in-field tests is to make sure that the network doesn't treat different CCs in different ways. 

If you are testing CCs that are already implemented in Kernel, you have pretty much a straight forward task. You don't need Pantheon's tunnel. You just need to bring up NTP servers (or use existing ones) and synch your machines up before running each individual experiment. 

However, if your tests include CCs that are only implemented in user-space (and most-likely over a UDP-based structure), you need to make sure that the network doesn't treat them in a different way compared to TCP-based CCs (UDP traffic vs TCP traffic).  That's where you need to use a tunnel-based solution like Pantheon to make the traffic of diff. CCs similar to each other (through encapsulating them). We did the same. We used Pantheon's tunneling to encapsulate the traffic. In addition to some changes on the main code, we added some modifications to make sure that our servers are always synched up to each other. 

If you elaborate on your test scenario in which you've observed your different results with/without Pantheon, we might be able to give you our thoughts on that but remember, these days the pandemic and the demand of remote-working will likely have an impact on your results especially if you're performing inter-continental experiments!

**Reply by ZoraNeverEatFish** [reply]
Hi @Soheil-ab 

Appreciate for your reply! Inspired by it, I run CCs in tunnel again and monitor the UDP port. I found that the UDP datagram had bad length. So the datagram would be split on IP layer which could cause accidental packet loss. After reducing packets' size, it works well now!

Best regards！

**Reply by Soheil-ab** [reply]
Hey @ZoraNeverEatFish 

Happy to hear that your issue is resolved!

**Comment by zilmarij** [comment]
Hi Soheil,

I am trying to run your framework as a part of my mini-project, on my system ( in VirtualBox, Ubuntu 18.04 with approximately 3GB of RAM, uname -r = 4.13.1-0521-learner) following the steps given in the README doc. I have got so far as to be able to do the build, but running the standalone version of Orca, I get the following:
python program terminated using kill -15
 orca-server-mahimahi: no process found
Terminated
...
$MAHIMAI_BASE 1 44444' : process exited with failure status 143 

As is obvious, it has something to do with Mahimahi. I cannot figure out the reason behind the 'no process found' error. 
Kindly help me out.

Thanks,
Zilmarij Iqbal

**Reply by Soheil-ab** [reply]
Hi @zilmarij 

First, I recommend you to start with an Ubuntu 16.04 as your base OS (and you may give it at least 4GB of RAM). Our current Kernel patch is based on Kernel v4.13. However, Ubuntu 18.04 and 20.04 are based on Kernel v5.X. So, it'll be safer to start with Ubuntu 16.04 to prevent some issues that you may face later. But, if you feel that everything is alright there, skip this recommendation! :)

Coming back to your question, plz give me the entire output. We usually clean up by killing various processes after a successful evaluation. So, those lines that you see can simply verify that everything is fine. 

If everything goes well, you can also check the "log" directory for the results of your evaluations (plz check actor.sh scripts for more detail).

**Reply by zilmarij** [reply]
Hello Soheil,

Maybe the ubuntu version is messing things up. I am trying with ubuntu 16 now. 
Even though you have mentioned about the path for venv in the README, I would still like to confirm if the venv needs to be setup inside the mahimahi or it could be placed outside.

**Reply by Soheil-ab** [reply]
@zilmarij, 

Regarding the venv path:
The current code expects to see the venv folder at home directory (_cd ~ ; mkdir venv;_). If you need to change its path, you should change the source code and update the expected venv path therein. 

This instruction is independent to the mahimahi installation or its path.

**Reply by zilmarij** [reply]
@Soheil-ab 

Thanks for the clarification.

Ubuntu 16 throws its own issues with pip3 and all, and the startup speed is just an icing-on-the-cake.
So what I did in case of ubuntu18 is that I booted into the 4.13.1-0521-learner from the grub through the BIOS (following the instructions in the README) and then tried to run the framework following the command specified.
What do you think could be going wrong?

**Reply by Soheil-ab** [reply]
@zilmarij ,

If you've already installed all required tools/packages and you don't see any errors when running the code, you are good to go!

**Reply by zilmarij** [reply]
@Soheil-ab 

I did it from scratch again, the same error shows.
Find attached the file containing the output of orca-standalone-emulation.
[out.txt](https://github.com/Soheil-ab/Orca/files/6344500/out.txt)

Please help me out.

**Reply by Soheil-ab** [reply]
@zilmarij ,
Re your file,

Lines 7-33 are general Tensorflow warnings (don't care them!). After Line 42, we terminate any existing processes, so what you see is normal. Now, results should be available @ rl-module/log directory. Checkout file named sum-<Your setting>.tr
